### PR TITLE
Fix a couple of warnings in C code.

### PR DESCRIPTION
**Pull request by shindere**
Cc @alainfrisch @dra27 @adrien-n @whitequark

The two important points here seem to be:

1. Warnings turned into errors on MinGW

2. Changes related to the _WIN32_WINNT macro

Comments welocme on everything in general and especially on these two
aspects.

### CODE REVIEW

File: config/Makefile.msvc [hunk 1]
```diff
@@ -95,7 +95,7 @@ SAFE_STRING=false
 BYTECC=cl -nologo -D_CRT_SECURE_NO_DEPRECATE -O2 -Gy- -MD
 
 ### Additional compile-time options for $(BYTECC).  (For static linking.)
-BYTECCCOMPOPTS=-DCAML_NAME_SPACE
+BYTECCCOMPOPTS=-DCAML_NAME_SPACE -D_WIN32_WINNT=0x0600
```
Comment (adrien-n) [inline config/Makefile.msvc]: So OCaml is now going to target >= Vista? Or am I wrong?

(unfortunately I don't have enough time to look at things properly now)

**Comment by shindere** [comment]
adrien-n (2016/11/17 07:01 -0800):

> adrien-n commented on this pull request.
> 
> > @@ -95,7 +95,7 @@ SAFE_STRING=false
> >  BYTECC=cl -nologo -D_CRT_SECURE_NO_DEPRECATE -O2 -Gy- -MD
> 
>  ### Additional compile-time options for $(BYTECC).  (For static linking.)
> -BYTECCCOMPOPTS=-DCAML_NAME_SPACE
> +BYTECCCOMPOPTS=-DCAML_NAME_SPACE -D_WIN32_WINNT=0x0600
> 
> So OCaml is now going to target >= Vista? Or am I wrong?
> 
> (unfortunately I don't have enough time to look at things properly
> now)

Well in my understanding it _did_ already target >= Vista, since
_WIN32_WINNT was set to 0x0600 in several places in the code.

The only change is that the macro is defined only once (for one given
build system).

**Comment by adrien-n** [comment]
On Thu, Nov 17, 2016, Sébastien Hinderer wrote:

> adrien-n (2016/11/17 07:01 -0800):
> 
> > adrien-n commented on this pull request.
> > 
> > > @@ -95,7 +95,7 @@ SAFE_STRING=false
> > >  BYTECC=cl -nologo -D_CRT_SECURE_NO_DEPRECATE -O2 -Gy- -MD
> > 
> >  ### Additional compile-time options for $(BYTECC).  (For static linking.)
> > -BYTECCCOMPOPTS=-DCAML_NAME_SPACE
> > +BYTECCCOMPOPTS=-DCAML_NAME_SPACE -D_WIN32_WINNT=0x0600
> > 
> > So OCaml is now going to target >= Vista? Or am I wrong?
> > 
> > (unfortunately I don't have enough time to look at things properly
> > now)
> 
> Well in my understanding it _did_ already target >= Vista, since
> _WIN32_WINNT was set to 0x0600 in several places in the code.
> 
> The only change is that the macro is defined only once (for one given
> build system).

OK, I saw it elsewhere but I didn't see it "officially" announced. Hence
my surprise. It also makes me wonder if it's actually wanted (I think it
makes more sense to drop <= XP but I don't mind much since I'm not
writing much support code nowadays).

Sorry for the hijack. :P

**Comment by shindere** [comment]
adrien-n (2016/11/17 07:11 -0800):

> OK, I saw it elsewhere but I didn't see it "officially" announced. Hence
> my surprise.

I understand. I am willing to update the PR with an official announce,
if that is needed.

> It also makes me wonder if it's actually wanted (I think it
> makes more sense to drop <= XP but I don't mind much since I'm not
> writing much support code nowadays).

Well, the point is that since some parts of the code defined it to
0X600, then they must have their reasons to do so. That's why this value
has been kept.

If a Windows expert says that a lower value is fine and won't break
anything we can certainly adopt a lower value.

> Sorry for the hijack. :P

No problem! You are very welcome IMO!

**Comment by adrien-n** [comment]
OK, back at home, I've checked the git history and afaict, the change was introduced by @dra27 when adding support for symlinks. It's been quite some time now so I'm wondering if anyone still tests, uses or distributes for XP. On that topic, note that mingw-w64 v5 was released roughly one month ago and v6 is very likely to not support XP officially (with the current pace of releases, v6 is due in one year or so).

File: config/Makefile.mingw [hunk 2]
```diff
@@ -101,7 +101,9 @@ SAFE_STRING=false
 BYTECC=$(TOOLPREF)gcc -O -mms-bitfields
 
 ### Additional compile-time options for $(BYTECC).  (For static linking.)
-BYTECCCOMPOPTS=-DCAML_NAME_SPACE -Wall -Wno-unused
+BYTECCCOMPOPTS=\
+  -DCAML_NAME_SPACE -D_WIN32_WINNT=0x0600 \
+  -Wall -Wno-unused -Werror
```
Comment (adrien-n) [inline config/Makefile.mingw]: I thought we weren't putting -Werror in releases but only for development. Has there been some discussion I haven't seen.

File: otherlibs/unix/access.c [hunk 3]
```diff
@@ -24,22 +24,23 @@
...
-#  define R_OK    4/* test for read permission */
-#  define W_OK    2/* test for write permission */
-#  define X_OK    4/* test for execute permission - not implemented in Win32 */
-#  define F_OK    0/* test for presence of file */
 # endif
 #endif
 
 static int access_permission_table[] = {
-  R_OK, W_OK, X_OK, F_OK
+  R_OK, W_OK,
```
Comment (adrien-n) [inline otherlibs/unix/access.c]: This is quite difficult to read.

Maybe making intermediate variables like "r_ok", "w_ok, "x_ok" and "f_ok" would be clearer; x_ok's value would depend on some #defines.

Another possibility would also be to split that first line ("R_OK, W_OK,") so that the various lines here get more consistent.

In any case, I think the whole thing is going to be difficult to make pretty but I think it's worth it: the variable value for X_OK was quite surprising.

File: otherlibs/win32graph/draw.c [hunk 4]
```diff
@@ -304,13 +304,22 @@ static value gr_draw_or_fill_arc(value vx, value vy, value vrx, value vry,
...
 
+/*
+ * The following function, caml_gr_show_bitmap, is broken:
+ * 1. AfficheBitmap which it calls is not exported (defined in dib.c)
+ * 2. The types of passed arguments does not match the expected ones:
+ *   void AfficheBitmap(char *filename,HDC hDC,int x,int y)
+ * 3. AfficheBitmap is called twice, which is probably a mistake
+ * Since this function is not used anywhere, it is disabled 
+ */
+#if 0
```
Comment (adrien-n) [inline otherlibs/win32graph/draw.c]: I'd remove this completely. Indeed, I don't know how this could be working and I don't like commented-out code.

Moreover you can remove AfficheBitmap(). And ChargerBitmap() because only AfficheBitmap() uses it. And it probably expands to most of dib.c if not all of it.

File: otherlibs/win32unix/unixsupport.h [hunk 5]
```diff
@@ -59,7 +59,9 @@ extern int win_CRT_fd_of_filedescr(value handle);
 
 extern void win32_maperr(DWORD errcode);
 extern value unix_error_of_code (int errcode);
-extern void unix_error (int errcode, char * cmdname, value arg);
+CAMLnoreturn_start
```
Comment (adrien-n) [inline otherlibs/win32unix/unixsupport.h:63]: Why is that change needed? From my quick searches in your top comment, I couldn't find a corresponding warning from the compiler.

**Comment by murmour** [comment]
So, the latest OCaml doesn't support Windows XP? Wow. This is big news. At [PrimeGeo](http://www.primegeo.ru/home.html), we still have customers who use XP for various reasons. I sometimes have to fix XP-specific stuff in our software product (a part of which is written in OCaml, by the way).

As of 2016, dropping support for XP is clearly premature. If a certain function from Vista is needed to support symlinks, then this function should be linked to the runtime dynamically (with an explicit call to `GetProcAddress`), and a call to `Unix.symlink` should throw an exception on XP.

**Comment by whitequark** [comment]
> As of 2016, dropping support for XP is clearly premature.

As of 2016, dropping support for a 15-year-old OS that has been completely unsupported in any way, shape or form for 2 years, and is full of publicly known vulnerabilities, is clearly sensible.

**Comment by murmour** [comment]
> As of 2016, dropping support for a 15-year-old OS that has been completely unsupported in any way, shape or form for 2 years, and is full of publicly known vulnerabilities, is clearly sensible. 

The fact is: this piece of crap of an OS still has a sizable market share, being deployed on [almost 10%](https://www.netmarketshare.com/operating-system-market-share.aspx) of all desktop computers. There are lots of legacy installations, even in enterprise. And you can't throw these users under the bus.

In case of OCaml, retaining support for XP should be trivial anyway. It only uses a single function from Vista, and I've offered a simple way to transform this static dependency into a dynamic one.

**Comment by adrien-n** [comment]
On Thu, Nov 17, 2016, Max Mouratov wrote:

> So, the latest OCaml doesn't support Windows XP? Wow. This is big news. At [PrimeGeo](http://www.primegeo.ru/home.html), we still have customers who use XP for various reasons. I sometimes have to fix XP-specific stuff in our software product (a part of which is in OCaml, by the way).

Don't take my word on it: I've never understood this part of Windows
well. I think it also depends on which msvcrt you ship.

There's also a weird impact of using msvcrt*.dll rather than the OS'
default CRT on GPL-licensed software installation (tbh, I bet that if it
ended being mentioned in court, noone would try to enforce because it
seems closer to bad wording and lack of foresight than anything else).

_But_ my point is rather than _as_ _far_ _as_ _I_ _can_ _tell_ (and
remember it's a thing I don't understand well), there should probably
have been an impact on machines with Windows XP. If noone noticed it's
maybe because noone tried because noone had a need for that support (or,
again², that I didn't understand something: please someone check!).

> As of 2016, dropping support for XP is clearly premature. If a certain function from Vista is needed to support symlinks, then this function should be linked to the runtime dynamically (with an explicit call to `GetProcAddress`), and a call to `Unix.symlink` should throw an exception on XP.

I strongly disagree on that. The main issue is that probably noone tests
the development version on XP and bugs can creep in easily; they won't
be noticed until release or even later on. There's also an issue of
manpower here.

I also strongly disagree on the API stuff. Symlinks are one example but
I much prefer threading APIs. XP lacks _many_ threading APIs,
particularly for locking; because of that there are some things you
cannot do on XP and some others you can only do at low performance.

I could go on and give other examples but of course some things are not
needed at the core of the runtime (symlinks) but you wouldn't remove
others (threading) anyway so why not take advantage of the
non-absolutely-required ones too.

Note that this is in Unix currently, not the whole codebase.

I encourage you to open a new ticket to debate this. Please provide some
research on the impact of the WIN32_WINNT behaviour on OS compatibility
and relationship with the various versions of msvcrt*.dll so that
everyone (including me) can understand the consequences.

@shindere, if this drags on, I think you can use 0x0502 (XP SP2) instead
of 0x0600 since it's used elsewhere and keep the current warning in
symlink.c (that is, if you either don't use -Werror or the right
-Werror=no-that-warning to make /that/ warning non-fatal). Another
possibility is to commit in two. Also, indeed, the current commit
propagates the define from otherlibs/unix/symlink.c to the whole C
codebase, including runtime.

**Comment by shindere** [comment]
adrien-n (2016/11/17 11:19 -0800):

> OK, back at home, I've checked the git history and afaict, the change
> was introduced by @dra27 when adding support for symlinks. It's been
> quite some time now so I'm wondering if anyone still tests, uses or
> distributes for XP.

Many thanks for your search, @adrien-n !

@dra27 : does your museum include XP machines?

**Comment by dra27** [comment]
Sorry to be late to the party here: Windows XP is, as far as I'm concerned, still meant to work and in fact I have a feeling that I recently restored Windows 2000 support for running OCaml executables (Cygwin hasn't supported Windows 2000 for ages, which makes the process of building OCaml on Windows 2000 far too much effort to bother testing).

At the moment, I'm not running any kind of CI for Windows XP, but that's meant to change in the next year with (hopefully) the introduction of continuous testing of various Windows versions.

If Windows XP is presently broken, please feel free to open a Mantis ticket (or submit a pull request). I'm quite happy to look into it - however, this is only happening while supporting XP remains trivial. Cygwin, for example, recently stopped supporting Windows XP because they wanted to start using "modern" (Vista+) APIs and working around Windows XP is too much work. OCaml's Windows XP support will likely end in the same way (and Adrien has given example reasons why that could happen).

As far as I remember, CreateSymbolicLink is dynamically linked - what might be broken here is the ability to compile the _compiler_ on Windows XP.

I agree with @adrien-n that having -DWIN32_WINNT=0x0502 and explicitly overriding it in files which need higher versions is a good idea.

File: config/Makefile.mingw [hunk 6]
```diff
@@ -168,7 +170,7 @@ SYSTEM=mingw
 NATIVECC=$(BYTECC)
 
 ### Additional compile-time options for $(NATIVECC).
-NATIVECCCOMPOPTS=-DCAML_NAME_SPACE -Wall -Wno-unused
+NATIVECCCOMPOPTS=$(BYTECCCOMPOPTS)
```
Comment (dra27) [inline config/Makefile.mingw]: This doesn't strike me as a good idea - if a user chooses to override `BYTECCCOMPOPTS` (e.g. either by editing `config/Makefile` or by specifying `make -f Makefile.nt BYTECCCOMPOPTS=... world.opt`) then it's strange that that would imply setting native code options too?

File: otherlibs/win32unix/symlink.c [hunk 7]
```diff
@@ -12,11 +12,6 @@
 /*                                                                        */
 /**************************************************************************/
 
-/*
- * Windows Vista functions enabled
- */
-#define _WIN32_WINNT 0x0600
-
```
Comment (dra27) [inline otherlibs/win32unix/symlink.c:19]: Don't forget to revert this if the -D switch is changed to 0x0502.

**Comment by shindere** [comment]
David Allsopp (2016/11/18 12:33 -0800):

> -/*
> - \* Windows Vista functions enabled
> - */
>   -#define _WIN32_WINNT 0x0600
>   -
> 
> Don't forget to revert this if the -D switch is changed to 0x0502.

Sorry about a likely dumb question but I really know nothing about
Windows.

What exactly would be the consequence of not reverting this changeif wi
set _WIN32_WINNT to 0x502 globally?

**Comment by shindere** [comment]
David Allsopp (2016/11/18 12:26 -0800):

> At the moment, I'm not running any kind of CI for Windows XP, but
> that's meant to change in the next year with (hopefully) the
> introduction of continuous testing of various Windows versions.

Having continuous integration would of course be great. Until this
happens, can you think of any possible way for me to have access to a
remote machine with XP installed so that I can verify that things work
correctly, also while doing changes?

**Comment by alainfrisch** [comment]
> As of 2016, dropping support for a 15-year-old OS that has been completely unsupported in any way, shape or form for 2 years, and is full of publicly known vulnerabilities, is clearly sensible.

I'm pretty sure that all OCaml developers would agree that using WinXP is a bit crazy, but they might need to support customers or users that are somehow forced to do it.  FWIW, we (LexiFi) have seen recently (last year) some big corporations still using WinXP.  Still, we have decided to stop supporting WinXP for our end-user application, but I'm not sure other companies selling OCaml software on Windows are on the same boat.  Anyway, I argue dropping support for WinXP (at least for runtime) even in 2016 cannot be done lightly.

**Comment by xavierleroy** [comment]
Let's not get carried away in this discussion.

Nobody seems to quite know what -D_WIN32_WINNT=xxxx actually does: does it just make visible (as in, declare in .h files) Win32 extensions introduced in or before version "xxxx"?  or does it prevent the application from running on Windows versions prior to "xxxx"?  My bet is the former, but it would be good to understand this point before arguing over this PR.

Regardless of the merits or demerits of Windows XP (*), I think it is the case that the core OCaml runtime system doesn't use any "recent" feature of the Win32, "recent" meaning "posterior to Win 2K" or maybe even "posterior to NT 3.5".  If this is true it would make sense to compile the runtime system without any "-D_WIN32_WINNT" option, or with one that selects a very old version of the Win32 API.

For the Unix library, I understand that symlink support needs Vista or later, so I'm OK for the corresponding _WIN32_WINNT define.  Whether it is done on the command line or as a #define in the incriminated source files makes little difference.  Choose whichever is easiest.


(*) Yes, I still occasionally see colleagues from industry presenting talks using laptops running XP.  This is even funnier when said colleagues work at companies from the military or security sectors.

**Comment by shindere** [comment]
The PR has been rebased and updated.

Regarding the _WIN32_WINNT macro: the commit remains conservative and
just undefs the macro before it gets defined, to get rid of the
warnings. A Mantis ticket has been opened to keep track of the issue:

https://caml.inria.fr/mantis/view.php?id=7418

@cakeplus : the Vista-specific function in symlink.c is already
dynamically loaded / linked.

@dra27 : the commit now defines NATIVECCCOMPOPTS explicitly again,
rather than using
NATIVECCCOMPOPTS=$(BYTECCCOMPOPTS)

@adrien-n:

- The -Werror flag has been dropped. Ideally one would want to have it
for developemnt versions but not for releases, as is done on Unix, but
that is not trivial to implement given the current state of the Windows
build system so it is probably simpler to leave this out until the build
systems have been merged.

- The patch in otherlibs/unix/access.c has been simplified

- Unused functions in win32graph have been removed

- The commit message has been update to explain why the change regarding
unix_error is needed:

3. The warnings about variables that might be used without having been
previously initialised occur because the C compiler does not realise
that unix_error never returns. So, fixing this warning does not require
any modification of the source files where it occurs. It rather requires
to declare that unix_error does not return on Windows, as is actually
already done on Unix.

**Comment by adrien-n** [comment]
Thanks for opening a topic dedicated to the issue of deciding whether to support Windows XP or not.

Sorry for the confusion about the compatibiliy. Indeed: merely defining WINNT_VERSION will not change anything since all it does is expose new APIs. In the current situation, the API is used through  the following code and will therefore not bomb out on XP:

     (LPFN_CREATESYMBOLICLINK)GetProcAddress(GetModuleHandle("kernel32"), "CreateSymbolicLinkA");

Were it not looked up dynamically as I had assumed at first, it would make the executables fail at startup.

(trying to find some time to review the most recent changes, not sure if I'll be able to complete that tonight)

**Comment by damiendoligez** [comment]
I have reviewed the PR and I'll merge it very soon unless somebody objects.

**Comment by adrien-n** [comment]
As far as I can tell, dib.c is still around in the latest version of the patch. However there are clearly unused functions, starting with Getfilesize().

I also don't understand why this isn't picked up by GCC: the function is static, unused and not visible from the outside (because there is no .h for it). There might be something to check there.

I have a minor comment about the comment in access.c. Currently it reads:
     +  /* There is no execute permission on Windows */ 
but I would prefer it to be expanded slightly in the vein of:
     +  /* Since there is no concept of execute permission on Windows, we fall back to the read permission */ 

Overall I'm very pleased with all your changes, thank you.

**Comment by shindere** [comment]
Many thanks for your feedbakc, @adrien-n.

> As far as I can tell, dib.c is still around in the latest version of
> the patch. However there are clearly unused functions, starting with
> Getfilesize().

The commit has been rebased and updated to remove the dib.c module
completely, since none of the functions it defines is actually used
outside of this module.

> I have a minor comment about the comment in access.c. Currently it reads:
>      +  /* There is no execute permission on Windows */ 
> but I would prefer it to be expanded slightly in the vein of:
>      +  /* Since there is no concept of execute permission on Windows,
>      we fall back to the read permission */

Good idea, it is indeed clearer like this. Done, thanks.

> Overall I'm very pleased with all your changes, thank you.

Thanks for telling! ;-)

**Comment by adrien-n** [comment]
On Sat, Nov 26, 2016, Sébastien Hinderer wrote:
> Many thanks for your feedbakc, @adrien-n.
> 
> > As far as I can tell, dib.c is still around in the latest version of
> > the patch. However there are clearly unused functions, starting with
> > Getfilesize().
> 
> The commit has been rebased and updated to remove the dib.c module
> completely, since none of the functions it defines is actually used
> outside of this module.

OK, thanks for the clarification. It seems Github's interface still
doesn't fully make sense to me. :)

**Comment by shindere** [comment]
adrien-n (2016/11/26 11:50 -0800):
> On Sat, Nov 26, 2016, Sébastien Hinderer wrote:
> > Many thanks for your feedbakc, @adrien-n.
> > 
> > > As far as I can tell, dib.c is still around in the latest version of
> > > the patch. However there are clearly unused functions, starting with
> > > Getfilesize().
> > 
> > The commit has been rebased and updated to remove the dib.c module
> > completely, since none of the functions it defines is actually used
> > outside of this module.
> 
> OK, thanks for the clarification. It seems Github's interface still
> doesn't fully make sense to me. :)

I mean that this update has been done as a follow-up to your comment, so
it has _just_ been updated! :)
Sorry for the confusion!

**Comment by adrien-n** [comment]
Ah, ok, actually this gives me some hope in my ability to use web
interfaces! It would have been nice that github sent an email
notification for that too however.

**Comment by shindere** [comment]
adrien-n (2016/11/26 11:54 -0800):
> Ah, ok, actually this gives me some hope in my ability to use web
> interfaces!

:)

> It would have been nice that github sent an email
> notification for that too however.

Ow, not sure it would have benn that nice: I push on github everyttime I
want to check on other machines so that would make you receive plenty of
e-mails, most of them being totally uninteresting!

That being said, I wouldn't be surprised if GitHub would provide a way
to receive this kind of notification, e.g. if you follow a user or
something like that.

**Comment by whitequark** [comment]
See "pull request pushes" on https://github.com/settings/notifications for enabling these notifications.

**Comment by shindere** [comment]
Thank you so much @whitequark !

**Comment by murmour** [comment]
I somehow missed recent activity here. Sorry.

It's great to know that `ocamlrun` still works on XP. Thanks, everyone, for taking the issue seriously.

@adrien-n 
> I also strongly disagree on the API stuff. Symlinks are one example but
I much prefer threading APIs. XP lacks many threading APIs,
particularly for locking; because of that there are some things you
cannot do on XP and some others you can only do at low performance.

Sure. I totally agree that maintaining support for XP only makes sense while it's trivial to do so. And if breaking compatibility is required to add new important features (like improved threading), then this topic (dropping platform support) should be discussed along with those changes.

**Comment by adrien-n** [comment]
Sebastien has opened a ticket dedicated to the issue of compatibility
with XP at https://caml.inria.fr/mantis/view.php?id=7418 . It's better
to keep the discussion on that topic centralized because it could make
many other discussions derail.

**Comment by dra27** [comment]
I think given how long that primitive has been disabled that none of the changes here fix any user-visible bugs?

**Comment by shindere** [comment]
David Allsopp (2016/11/27 05:52 -0800):
> I think given how long that primitive has been disabled that none of
> the changes here fix any user-visible bugs?

Which primitive are you talking about, actually?

**Comment by dra27** [comment]
@shindere: `caml_gr_show_bitmap`

**Comment by dra27** [comment]
Oh, hang on - I see looking closer that although it's declared CAMLprim, it's not in fact a full primitive is it - so only C code which linked against graphics might have used that which seems **even** less likely!

**Comment by shindere** [comment]
David Allsopp (2016/11/28 06:07 -0800):
> Oh, hang on - I see looking closer that although it's declared
> CAMLprim, it's not in fact a full primitive is it - so only C code
> which linked against graphics might have used that which seems
> **even** less likely!

Indeed. Will try to check for it though. Thanks.

>

**Comment by shindere** [comment]
David Allsopp (2016/11/28 06:05 -0800):
> @shindere: `caml_gr_show_bitmap`

OK. I checked on Google and the only results it provided on its first
page were from the OCaml sources themselves.

**Comment by shindere** [comment]
Just rebased this PR on the latest trunk.

Does it seem ready for merging, now?

File: otherlibs/win32unix/unixsupport.h (hunk 5 above)
Comment (dra27) [reply otherlibs/win32unix/unixsupport.h:63]: It's present in the unix version - however, there is also present around `uerror` - should it be here too?

**Comment by shindere** [comment]
David Allsopp (2016/12/06 06:13 -0800):
> > @@ -59,7 +59,9 @@ extern int win_CRT_fd_of_filedescr(value handle);
>  
>  extern void win32_maperr(DWORD errcode);
>  extern value unix_error_of_code (int errcode);
> -extern void unix_error (int errcode, char * cmdname, value arg);
> +CAMLnoreturn_start
> 
> It's present in the unix version - however, there is also present
> around `uerror` - should it be there too?

Yeah why not, thanks!
The PR has been updated.

**Comment by dra27** [comment]
This looks good to me - a related thought once it's merged is whether we would like to alter the Travis/AppVeyor scripts so that they compile with C compiler (and OCaml compiler warnings now, I think?) warnings turned into errors?

File: otherlibs/win32unix/unixsupport.h [hunk 8]
```diff
@@ -59,8 +59,15 @@ extern int win_CRT_fd_of_filedescr(value handle);
...
 extern void win32_maperr(DWORD errcode);
 extern value unix_error_of_code (int errcode);
-extern void unix_error (int errcode, char * cmdname, value arg);
+
+CAMLnoreturn_start
+extern void unix_error (int errcode, char * cmdname, value arg)
+CAMLnoreturn_end;
+
+CAMLnoreturn_start
 extern void uerror (char * cmdname, value arg);
```
Comment (dra27) [inline otherlibs/win32unix/unixsupport.h]: I don't think there should be a semicolon at the end of this line?

**Comment by shindere** [comment]
David Allsopp (2016/12/06 07:28 -0800):
> This looks good to me - a related thought once it's merged is whether
> we would like to alter the Travis/AppVeyor scripts so that they
> compile with C compiler warnings turned into errors?

Not sure whether the build system in its present state lets one override
CFLAGS when running make.

Of course if appveyor could be given a config/Makefile that is the one
from the repo but with these flags added, that would be great.

This will all be solved once autoconf is in place because then
warnings-as-errors can be turned on automatically on dev versions as
they are currently on Unix.

**Comment by shindere** [comment]
David Allsopp (2016/12/06 07:34 -0800):
> I don't think there should be a semicolon at the end of this line?

Yeah absolutely you are right of course!
Thanks for being so careful and sorry for having been too lazy to try to
compile this once more.

**Comment by dra27** [comment]
It's not a compiler error at the moment because of how the `#define`s are set-up, just a philosophical one!

**Comment by shindere** [comment]
David Allsopp (2016/12/06 07:48 -0800):
> It's not a compiler error at the moment because of how the `#define`s
> are set-up, just a philosophical one!

I think due to the way the macros are defined it would have caused a
syntax error, actualy.
